contracts/ControllerCurveUniV3Strat.sol

```diff
@@ -0,0 +1,327 @@
+    }
+
+    function setUseNative(bool _useNative) external onlyAdmin {
+        useNative = _useNative;
+    }
+
+    function setTokenToTokenSwapFee(address _tokenA, address _tokenB, uint24 _fee) external onlyAdmin {
+        require(_tokenA != address(0), "!ZeroAddress tokenA");
+        require(_tokenB != address(0), "!ZeroAddress tokenB");
+        require(_fee >= 0, "Fee can't be negative");
+
+        tokenToTokenSwapFee[_tokenA][_tokenB] = _fee;
+    }
+
+    function _deposit() internal override {
+        uint _wantBal = wantBalance();
+
+        if (_wantBal > 0) {
+            _addLiquidity(_wantBal);
+        }
+
+        uint _wantCRVBalance = wantCRVBalance();
+
+        if (_wantCRVBalance > 0) {
+            crvToken.safeApprove(address(gauge), _wantCRVBalance);
+            gauge.deposit(_wantCRVBalance);
+        }
+    }
+
+    function _addLiquidity(uint _wantBal) internal {
+        uint _expected = _wantToWantCrvDoubleCheck(_wantBal, true);
+
+        if (poolSize == 2) {
+            uint[2] memory _amounts;
+
+            _amounts[uint(uint128(tokenIndex))] = _wantBal;
+
+            if (useNative) {
+                IWNative(address(want)).withdraw(_wantBal);
+                ICurvePool(pool).add_liquidity{value: address(this).balance}(_amounts, _expected);
+            } else {
+                want.safeApprove(pool, _wantBal);
+                ICurvePool(pool).add_liquidity(_amounts, _expected, true);
+            }
+        } else if (poolSize == 4) {
+            uint[4] memory _amounts;
+
+            _amounts[uint(uint128(tokenIndex))] = _wantBal;
+
+            want.safeApprove(pool, _wantBal);
+            ICurvePool(pool).add_liquidity(_amounts, _expected);
+        }
+    }
+
+    function _withdraw(uint _amount) internal override returns (uint) {
+        uint _balance = wantBalance();
+
+        _withdrawFromPool(
+            _wantToWantCrvDoubleCheck(_amount - _balance, false)
+        );
+
+        uint _withdrawn = wantBalance() - _balance;
+
+        return (_withdrawn > _amount) ? _amount : _withdrawn;
+    }
+
+    function _withdrawAll() internal override returns (uint) {
+        uint _balance = wantBalance();
+
+        _withdrawFromPool(balanceOfPool());
+
+        return wantBalance() - _balance;
+    }
+
+    function _withdrawFromPool(uint _wantCrvAmount) internal {
+        // Remove staked from gauge
+        gauge.withdraw(_wantCrvAmount);
+
+        // remove_liquidity
+        uint _balance = wantCRVBalance();
+        uint _expected = _wantCrvToWantDoubleCheck(_balance);
+
+        require(_expected > 0, "remove_liquidity expected = 0");
+
+        if (address(pool) != address(swapPool)) {
+            crvToken.safeApprove(pool, _balance);
+        }
+
+        if (useNative) {
+            ICurvePool(pool).remove_liquidity_one_coin(_balance, tokenIndex, _expected);
+
+            uint _nativeBalance = address(this).balance;
+
+            if (_nativeBalance > 0) {
```

Comment:
here instead of an `if` i'ld put a `require(_nativeBalance >= _expected)`  or just nothing... we should "trust" in the received amount.